mlir/lib/Catalyst/Transforms/DetensorizeSCFPass.cpp

```diff
@@ -0,0 +1,354 @@
+            rewriter.setInsertionPoint(condOp);
+            for (const auto &it : llvm::enumerate(condOpArgs)) {
+                Value condOpArg = it.value();
+                if (isScalarTensor(condOpArg)) {
+                    Value value = rewriter.create<tensor::ExtractOp>(condOp->getLoc(), condOpArg,
+                                                                     ValueRange{});
+                    newCondOpArgs.push_back(value);
+                }
+                else {
+                    newCondOpArgs.emplace_back(condOpArg);
+                }
+            }
+            rewriter.replaceOpWithNewOp<scf::ConditionOp>(condOp, condOp.getCondition(),
+                                                          newCondOpArgs);
+        }
+
+        // 5. Extract scalar tensor elements and detensorize terminator
+        {
+            scf::YieldOp clonedYieldOp = cast<scf::YieldOp>(newAfterBlock.getTerminator());
+            SmallVector<Value> newYieldOperands = clonedYieldOp.getOperands();
+            OpBuilder::InsertionGuard g(rewriter);
+            rewriter.setInsertionPoint(clonedYieldOp);
+            for (std::size_t index : newIterOperandsIndices) {
+                newYieldOperands[index] = rewriter.create<tensor::ExtractOp>(
+                    clonedYieldOp->getLoc(), clonedYieldOp->getOperand(index), ValueRange{});
+            }
+            rewriter.create<scf::YieldOp>(newWhileOp.getLoc(), newYieldOperands);
+            rewriter.eraseOp(clonedYieldOp);
+        }
+
+        // 6. Retensorize results after for loop
+        {
+            OpBuilder::InsertionGuard g(rewriter);
+            rewriter.setInsertionPointAfter(whileOp);
+            for (std::size_t index : newIterOperandsIndices) {
+                Value for_result = newWhileOp->getResult(index);
+                Value value = rewriter.create<tensor::FromElementsOp>(
+                    newWhileOp->getLoc(), RankedTensorType::get({}, for_result.getType()),
+                    for_result);
+                rewriter.replaceUsesWithIf(whileOp->getResult(index), value, [&](OpOperand &op) {
+                    return !isa<tensor::FromElementsOp>(op.getOwner());
+                });
+            }
+        }
+
+        rewriter.replaceOp(whileOp, newWhileOp);
+        return success();
+    }
+};
+
+} // namespace
+
+namespace catalyst {
+#define GEN_PASS_DEF_DETENSORIZESCFPASS
+#include "Catalyst/Transforms/Passes.h.inc"
+
+struct DetensorizeSCFPass : public impl::DetensorizeSCFPassBase<DetensorizeSCFPass> {
+    using impl::DetensorizeSCFPassBase<DetensorizeSCFPass>::DetensorizeSCFPassBase;
+    void runOnOperation() override
+    {
+        MLIRContext *context = &getContext();
+        RewritePatternSet patterns(context);
+        patterns.add<DetensorizeForOp>(patterns.getContext());
+        patterns.add<DetensorizeIfOp>(patterns.getContext());
+        patterns.add<DetensorizeWhileOp>(patterns.getContext());
```

Comment:
Isnt' that just `context`? 😅 
```suggestion
        patterns.add<DetensorizeForOp>(context);
        patterns.add<DetensorizeIfOp>(context);
        patterns.add<DetensorizeWhileOp>(context);
```